bin/pasta_prepare_evaluation.py

```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
+            return email.message_from_bytes(msg)[field]
+        else:
+            return None
+
+    dd1 = dd.read_csv(config.f_denorm_responses, blocksize=1e9, dtype={"idx ": "int32", "patch_id ": "category",
+                                                                       "responses.resp_msg_id": "category",
+                                                                       "responses.parent": "category"})
+
+    dd1 = dd1.set_index(['idx'])
+
+    dd2 = dd.read_csv(config.f_denorm_upstream, blocksize=1e9, dtype={"idx ": "int32", "patch_id ": "category",
+                                                                      "upstream": "category"})
+
+    dd2 = dd2.set_index(['idx'])
+
+    df_dask_final = dd.merge(dd1, dd2, left_index=True, right_index=True, how='left') \
+        .drop(['patch_id_y'], axis=1) \
+        .reset_index(drop=True) \
+        .rename(columns={"patch_id_x": "patch_id"})
+
+    df_dask_final.to_csv("df_dask_final.csv", single_file=True)
+
+    final = dd.read_csv("df_dask_final.csv", blocksize=50e7, dtype={"idx ": "int32", "patch_id ": "category",
+                                                                    "responses.resp_msg_id": "category",
+                                                                    "responses.parent": "category",
+                                                                    "upstream": "category"}).drop('Unnamed: 0', axis=1)
+
+    print("Final shape with possible duplicate rows{}".format(final.shape))
+    final.drop_duplicates(inplace=True)
+
+    # Convert to pandas
+    df_pd_final = final.compute()
+
+    # Remove rows with no patch and other infos
+    index_names = df_pd_final[(df_pd_final['patch_id'] == '_') & (df_pd_final['responses.message'].isna()) &
+                              (df_pd_final['upstream'].isna())].index
+    df_pd_final.drop(index_names, inplace=True)
+
+    print("Final shape after removing duplicates {}".format(final.shape))
+
+    # df_pd_final.to_csv(config.f_merged_responses_upstream, index=False)
+    # print("Finished writing de-duplicated pandas merged dataframe to disk")
+
+    final = dd.from_pandas(df_pd_final, npartitions=20)
+
+    final['responses.message'] = final['responses.message'].map(try_literal_eval)
+
+    final.reset_index().rename(columns={'index': 'idx'}).compute()
+
+    final['response_author'] = final['responses.message'].map(lambda x: _get_message_field(x, 'from'),
+                                                              meta=pd.Series([], dtype=object, name='x'))
+
+    log.info("Unique response authors {}".format(final['response_author'].nunique().compute(num_workers=20)))
+
+    final.to_csv(config.f_responses_authors, single_file=True)
+
+
+def _is_response_from_bot(message_id):
+    message = _repo.mbox.get_messages(message_id)[0]
+    email = email_get_from(message)[1].lower()
+    bots = ['tip-bot2@linutronix.de', 'tipbot@zytor.com', 'noreply@ciplatform.org', 'syzbot',
```

Comment:
Why didn't you use / improve the stuff that we have in LinuxMailcharacteristics.py?

Reply:
I had to extract this part out, as the method _is_from_bot in LinuxMailcharacteristics is tightly coupled with the class. This prevents me from using it when a message is not a patch and things like is_patch, or _is_next are irrelevant for a response. This is not how it should finally look, and I look forward to your thoughts how to combine this best. I have to also admit, I do not fully understand all functionality of LinuxMailcharacteristics regarding the domain.

Looking forward to your remarks here.

Reply:
Yes, that's correct. At the moment it's tightly coupled. @ditzlike had a similar issue -- she wanted to use MailCharacteristics to assign a mail to the latest kernel version that released at that point in time. In this case, we had tight coupling, but in fact there's no need to instantiate a MailCharacteristics instance. In her case, it was possible to factor out _patch_get_version(): ref. b429ec028860fbe095cbc9f8d741abe264f58acf .

Maybe it's (somehow) possible to convert _is_from_bot() to a staticmethod.

Reply:
I used LinuxMailCharacteristics and improved the _is_from_bot method in two new patches